sled-agent/src/instance.rs

```diff
@@ -186,9 +198,117 @@ struct PropolisSetup {
     running_zone: RunningZone,
 }
 
-struct InstanceInner {
+// Requests that can be made of instances
+enum InstanceRequest {
+    RequestZoneBundle {
+        tx: oneshot::Sender<Result<ZoneBundleMetadata, BundleError>>,
+    },
+    CurrentState {
+        tx: oneshot::Sender<SledInstanceState>,
+    },
+    PutState {
+        state: crate::params::InstanceStateRequested,
+        tx: oneshot::Sender<Result<InstancePutStateResponse, ManagerError>>,
+    },
+    PutMigrationIds {
+        old_runtime: InstanceRuntimeState,
+        migration_ids: Option<InstanceMigrationSourceParams>,
+        tx: oneshot::Sender<Result<SledInstanceState, ManagerError>>,
+    },
+    Terminate {
+        tx: oneshot::Sender<Result<InstanceUnregisterResponse, ManagerError>>,
+    },
+    IssueSnapshotRequest {
+        disk_id: Uuid,
+        snapshot_id: Uuid,
+        tx: oneshot::Sender<Result<(), ManagerError>>,
+    },
+    AddExternalIp {
+        ip: InstanceExternalIpBody,
+        tx: oneshot::Sender<Result<(), ManagerError>>,
+    },
+    DeleteExternalIp {
+        ip: InstanceExternalIpBody,
+        tx: oneshot::Sender<Result<(), ManagerError>>,
+    },
+}
+
+// A small task which tracks the state of the instance, by constantly querying
+// the state of Propolis for updates.
+//
+// This task communicates with the "InstanceRunner" task to report status.
+struct InstanceMonitorRunner {
+    client: Arc<PropolisClient>,
+    tx_monitor: mpsc::Sender<InstanceMonitorRequest>,
+}
+
+impl InstanceMonitorRunner {
+    async fn run(self) -> Result<(), anyhow::Error> {
```

Comment:
See "monitor_state_task" below for how this was modified.

It's similar to the original, but rather than accessing the "instance" state via `Arc`, we chat with it on the `tx_monitor/rx_monitor` channel.